### PR TITLE
docs: Update start/install section

### LINKED ISSUE #13408: Improve documentation around running Envoy in Docker container
## description

a conversation came up in the slack channel around running Envoy in (non-root) docker containers.

i think most of the relevant information is there (https://www.envoyproxy.io/docs/envoy/latest/start/start#using-the-envoy-docker-image) but the section could do with being broken up a bit and possibly moved

### LINKED ISSUE #13491: brew install envoy
Hi, thanks for this great project! Just wanted to let you know that Envoy is now available on Homebrew, so Mac users can install it with:

```sh
brew install envoy
```

Thanks to @rebello95 and others involved in the `-Wrange-loop-analysis` work for making this possible.

Ref: https://github.com/Homebrew/homebrew-core/pull/61858

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs: Update start/install section
Additional Description:

This PR will

- split start page into separate pages and reorganize install section (see discussion https://github.com/envoyproxy/envoy/pull/13439#discussion_r502518032)
- improve docker information (#13408)
- add an install page (#13461)

Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue] Fix #13408 Fix #13491  touch #13461 touch #13439 
[Optional Deprecated:]

### CODE REVIEW

**Comment by phlax** [comment]
after doing a bit of initial moving around i think it looks good:

https://389760-65214191-gh.circle-artifacts.com/0/generated/docs/start/start.html

im wondering about the `ref_configs` page:

https://389760-65214191-gh.circle-artifacts.com/0/generated/docs/start/ref_configs.html

specifically, im wondering whether the sandboxes are kinda ref configs, and we should move the "Configuration generator" into tools

**Comment by mattklein123** [comment]
> specifically, im wondering whether the sandboxes are kinda ref configs, and we should move the "Configuration generator" into tools

Yeah I was going to discuss this with you. IMO we should merge configs/examples in the repo, and probably just move the config generator into examples? I think this would make more sense since what we have in configs at this point is really just another type of example.

Is this ready for review or still WIP?

**Comment by phlax** [comment]
> Is this ready for review or still WIP?

still WIP im afraid - i have almost finished the install and docker pages:

- https://390000-65214191-gh.circle-artifacts.com/0/generated/docs/start/install.html
- https://390000-65214191-gh.circle-artifacts.com/0/generated/docs/start/docker.html

so  its worth having a look over them and the overall layout so far, but im still working on the rest

**Comment by phlax** [comment]
im minded to land the "default configuration for docker" section as is for now

https://390000-65214191-gh.circle-artifacts.com/0/generated/docs/start/docker.html#running-envoy-docker-with-the-default-configuration

and tidy up after (re #13497)

**Comment by phlax** [comment]
> Yeah I was going to discuss this with you. IMO we should merge configs/examples in the repo, and probably just move the config generator into examples? I think this would make more sense since what we have in configs at this point is really just another type of example.

i had thought to move it to tools - let me update and we can figure from there

**Comment by phlax** [comment]
i see now - you mean move the actual ref configs to sandboxes - yep, i think thats a good idea.

i had only started reading the "config generator" part, which i think can go to tools

**Comment by phlax** [comment]
hmm, this is what the ref_configs point to https://390000-65214191-gh.circle-artifacts.com/0/generated/docs/intro/deployment_types/deployment_types.html

i think we might want to look at the `intro` section after, but for now i have just remove the page from the `start` section and redirected here (i can do something different)

**Comment by phlax** [comment]
@mattklein123 this is the last build on circleci https://390172-65214191-gh.circle-artifacts.com/0/generated/docs/start/start.html

the `install` and `docker` sections are almost complete, and im half way through the `quick-start` (text will be readded)

probs worth looking over them as there are a lot of changes - trying to figure how i can add doc artifacts asap...

File: docs/root/intro/deployment_types/double_proxy.rst [hunk 1]
```diff
@@ -22,5 +22,4 @@ Configuration template
 ^^^^^^^^^^^^^^^^^^^^^^
 
 The source distribution includes an example double proxy configuration that is very similar to
-the version that Lyft runs in production. See :ref:`here <install_ref_configs>` for more
-information.
+the version that Lyft runs in production.
```
Comment (mattklein123) [inline docs/root/intro/deployment_types/double_proxy.rst:25]: This isn't actually true anymore. Can you remove this statement here and elsewhere as long as you are in here? I would just make this more generic in terms of showing a rudimentary templated configuration system or something like that.

File: docs/root/start/start.rst [hunk 2]
```diff
@@ -12,222 +12,12 @@ These examples use the :ref:`v3 Envoy API <envoy_api_reference>`, but use only t
...
--------------------------
-
 .. toctree::
     :maxdepth: 2
 
+    install
+    docker
+    quick-start
+    tools/tools
+    sandboxes/index
```
Comment (mattklein123) [inline docs/root/start/start.rst:22]: Personally I would swap the order on these?

File: docs/root/start/tools/config_generator.rst [hunk 3]
```diff
@@ -1,18 +1,3 @@
...
-The source distribution includes a set of example configuration templates for each of the three
-major Envoy deployment types:
-
-* :ref:`Service to service <deployment_type_service_to_service>`
-* :ref:`Front proxy <deployment_type_front_proxy>`
-* :ref:`Double proxy <deployment_type_double_proxy>`
-
-The goal of this set of example configurations is to demonstrate the full capabilities of Envoy in
-a complex deployment. All features will not be applicable to all use cases. For full documentation
-see the :ref:`configuration reference <config>`.
```
Comment (mattklein123) [inline docs/root/start/tools/config_generator.rst:15]: Per my other comment can you remove "Lyft" from below and make it more generic?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Overall this is amazing! I love it. Left a few small comments. LMK when it's ready for full review.

/wait

**Comment by phlax** [comment]
@mattklein123 ive created a PR to deal with the Lyft references separately #13559

File: docs/root/start/start.rst (hunk 2 above)
Comment (phlax) [reply docs/root/start/start.rst:22]: agreed, ill update...

Comment (phlax) [reply docs/root/start/start.rst]: i  also swapped docker <> quick-start

File: docs/root/start/start.rst [hunk 4]
```diff
@@ -12,222 +12,20 @@ These examples use the :ref:`v3 Envoy API <envoy_api_reference>`, but use only t
...
-
-
-Sandboxes
----------
-
-We've created a number of sandboxes using Docker Compose that set up different
-environments to test out Envoy's features and show sample configurations. As we
-gauge peoples' interests we will add more sandboxes demonstrating different
-features. The following sandboxes are available:
+    sandboxes/index
```
Comment (phlax) [inline docs/root/start/start.rst:25]: ive  put the sandboxes in a separate toc tree, because there are quite a few and its less helpful having the steps printed out in the tree.

this means there are 3 tocs ~joined togetther, and a small visual inconsistency between them

im wondering if as we have more of these we might  want to move the sandboxes to the bottom, which would also fix the toc problem

Comment (mattklein123) [reply docs/root/start/start.rst:25]: Makes sense to me. Let me know when this is ready for final review!

/wait

**Comment by phlax** [comment]
> Let me know when this is ready for final review!

its very close - i would like to do a littlle work on the remaining pages, but i have almost finished the quick-start

i have worked on the assumption that we ~fix the issue with the admin listening address - i probs also need to update here if we switch from google proxy also

File: docs/root/start/start.rst (hunk 4 above)
Comment (phlax) [reply docs/root/start/start.rst:27]: i moved the sandboxes to the bottom as we already have another incoming sandboxes

File: docs/root/start/_include/dockerhub-images.csv [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+`envoyproxy/envoy-alpine <https://hub.docker.com/r/envoyproxy/envoy-alpine/tags/>`_,Release binary with symbols stripped on top of a **glibc** alpine base.,v1.16-latest,,,
+`envoyproxy/envoy-debug <https://hub.docker.com/r/envoyproxy/envoy-debug/tags/>`_,Release binary with debug symbols on top of an Ubuntu Bionic base.,v1.16-latest,v1.16-latest,,
+`envoyproxy/envoy-dev <https://hub.docker.com/r/envoyproxy/envoy-dev/tags/>`_,Release binary with symbols stripped on top of an Ubuntu Bionic base.,,,latest,latest
+`envoyproxy/envoy-alpine-dev <https://hub.docker.com/r/envoyproxy/envoy-alpine-dev/tags/>`_,Release binary with symbols stripped on top of a **glibc** alpine base.,,,latest,
+`envoyproxy/envoy-debug-dev <https://hub.docker.com/r/envoyproxy/envoy-debug-dev/tags/>`_,Release binary with debug symbols on top of an Ubuntu Bionic base.,,,latest,latest
+`envoyproxy/envoy-build-ubuntu <https://hub.docker.com/r/envoyproxy/envoy-debug-dev/tags/>`_,Build image which includes tools for building multi-arch Envoy and containers.,,,latest(?),latest(?)
```
Comment (phlax) [inline docs/root/start/_include/dockerhub-images.csv:9]: not sure what to do with `latest(?)`

File: docs/root/start/building.rst [hunk 6]
```diff
@@ -28,62 +29,6 @@ for more information on performing manual builds.
...
-* `envoyproxy/envoy <https://hub.docker.com/r/envoyproxy/envoy/tags/>`_: Release binary with
-  symbols stripped on top of an Ubuntu Bionic base.
-* `envoyproxy/envoy-debug <https://hub.docker.com/r/envoyproxy/envoy-debug/tags/>`_: Release
-  binary with debug symbols on top of an Ubuntu Bionic base.
-* `envoyproxy/envoy-alpine <https://hub.docker.com/r/envoyproxy/envoy-alpine/tags/>`_: Release
-  binary with symbols stripped on top of a **glibc** alpine base.
-
-.. note::
-
-  In the above repositories, we tag a *vX.Y-latest* image for each security/stable release line.
```
Comment (phlax) [inline docs/root/start/building.rst:48]: readd this line in the note under table

**Comment by phlax** [comment]
@mattklein123 i think this is ready for review

i have a few minor changes left to make, so will update a little

i havent looked to much at the tools section, just left what was there and added the configuration_generator tool page

the build page probs needs some work, but i think it can be done in the context of #13431

**Comment by phlax** [comment]
just to mention, the docker images table comes from a csv file

this could theoretically be automatically generated - but for now it hardcodes the last stable - ie `v1.16-latest` we will need to update one way or another when last stable version changes

**Comment by phlax** [comment]
also worth mentioning, i would have linked to the dockerhub page of the images in the table, except for that you have to know the image sha digest to do that.

this could be figured out programatically, but it would mean the docs would necessarily have to be published after docker images and adds too many layers of complexity

so the only links are to the main `image/tags` page in the docker images table.

**Comment by phlax** [comment]
the other thing im thinking to mention - which is as much i think an issue with the rest of the docs/READMEs etc

i have used `master` to  refer to the branch used as a development head.

reluctantly, this contravenes the community policy here https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#inclusive-language-policy

i feel this is ok (the policy does the same in the following paragraph https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#breaking-change-policy), but more i think its a necessary evil for now as that is "precisely" the current name of the branch

i am very excited about the shift in the tech community to address these terminology issues - but i think  we have to wait until the  github `master` -> `main` shift before we can document it that way or risk confusing docs/commands/users

File: docs/root/start/_include/dockerhub-images.csv (hunk 5 above)
Comment (mattklein123) [reply docs/root/start/_include/dockerhub-images.csv:9]: So per the other issue, I'm fine with adding latest for this image. @lizan?

Stepping back though, WDYT about moving the documentation of this container to the building section? IMO it doesn't really fit in this section?

File: docs/root/start/_include/dockerhub-images.csv [hunk 7]
```diff
@@ -0,0 +1,9 @@
+,,stable,stable,master,master
+Docker image,description,amd64,arm64,amd64,arm64
+`envoyproxy/envoy <https://hub.docker.com/r/envoyproxy/envoy/tags/>`_,Release binary with symbols stripped on top of an Ubuntu Bionic base.,v1.16-latest,v1.16-latest,,
```
Comment (mattklein123) [inline docs/root/start/_include/dockerhub-images.csv:3]: WDYT about documenting the release vs. dev distinction here in the table also to make it super clear? As part of the description? (Basically moving or duplicating the below note "In the above dev repositories, the latest tag points to a container including the last Envoy build on master that passed tests." into here?)

Also, can you ping me on Slack with the text you want in the description for each repo and I will update it?

File: docs/root/start/start.rst [hunk 8]
```diff
@@ -12,222 +12,16 @@ These examples use the :ref:`v3 Envoy API <envoy_api_reference>`, but use only t
...
--------------------------
-
 .. toctree::
-    :maxdepth: 2
+    :maxdepth: 3
 
+    install
+    quick-start
+    docker
+    tools/tools
```
Comment (mattklein123) [inline docs/root/start/start.rst:21]: In looking at this now, WDYT about moving tools under operations and administration? I think it's a bit in the weeds for the getting started guide?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks @phlax this is really fantastic work. I have a few comments and then I suggest we ship and iterate as this is a gigantic improvement!!!

/wait

File: docs/root/start/building/local_docker_build.rst [hunk 9]
```diff
@@ -19,6 +20,12 @@ That command will take some time to run because it is compiling an Envoy binary
 
 For more information on building and different build targets, please refer to :repo:`ci/README.md`.
 
+.. warning::
```
Comment (phlax) [inline docs/root/start/building/local_docker_build.rst:23]: i think these instructions need a much stronger health warning

im uncomfortable landing this

Comment (mattklein123) [reply docs/root/start/building/local_docker_build.rst:23]: What do you suggest?

TBH building Envoy is an advanced topic and I wonder if we should remove it from the getting started guide entirely.

File: docs/root/start/start.rst (hunk 8 above)
Comment (phlax) [reply docs/root/start/start.rst:21]: wfm

File: docs/root/start/building/local_docker_build.rst (hunk 9 above)
Comment (phlax) [reply docs/root/start/building/local_docker_build.rst:23]: The problem is that i ended up on this page very quickly. I really dont like these docs, but  if we remove google will just turn up the old versions and people will still follow, or  it wont, and there are no docs for building a docker image.

This was my route into envoy

- check for official packages - none
- check for docker images - no latest, unclear versioning, no info page, unclear Dockerfile source
- ok, ill build my own (as i commonly do if i dont trust existing binary channels)
- land on this page, follow instructions, both hose my (resource-constrained) machine for different reasons, swear a lot
- read source code of run_envoy_docker and swear a lot more

at this point i wrongly concluded that the project wasnt mature and/or well supported, and certainly wasnt  well documented.

thankfully, i opened a PR with the unroot as that is what i had added in my own recipe, so i gave the  project further consideration

tldr - i think it cant stay there as it is, but at this point i think building your own containers should not be regarded as an advanced task, because i think it is often necessary

File: docs/root/start/_include/dockerhub-images.csv (hunk 5 above)
Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:9]: > Stepping back though, WDYT about moving the documentation of this container to the building section? IMO it doesn't really fit in this section?

not sure if i agree entirely. The existing docs spread info about images in a way that was really hard to grok. Ive tried to put info about all active images in one place - to be authoritative.

i would have separated the table with subheadings, but after quite a bit of trying i wasnt able to get sphinx to do colspans (it doesnt seem to be documented that it can, but it is part of the rst spec afaict)

ultimately moving it doesnt resolve the `latest(?)` issue it just kicks it to another page

File: docs/root/start/_include/dockerhub-images.csv (hunk 7 above)
Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:3]: > WDYT about documenting the release vs. dev distinction 

that its a bad idea.

afaict they are the same images (built with the same recipe) with the same binaries inside (built with same compiler) just at different commit points

they dont do anything different to each other

Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:3]: in fact the note about master being release-quality kinda says as much

File: docs/root/start/building/local_docker_build.rst (hunk 9 above)
Comment (phlax) [reply docs/root/start/building/local_docker_build.rst:23]: > What do you suggest?

addressing the underlying issues - eg the issues/prs that i opened.

a couple more things also i think - separating docs and docker build containers from main development/build container.

The docker build container has no need for most of the compile stuff but does need qemu, and host/root privileges. It could be very small and documented exactly what it does and how it works

The docs build doesnt need 90% of what is in the dev/build image. The current setup requires a 5GB image to run sphinx and protoc

Lastly, i guess there is a good reason for statically linking all of the compiled llvm binaries - its beyond my compiler knowledge to understand why - but that afaict is what makes that image so large

Comment (phlax) [reply docs/root/start/building/local_docker_build.rst:23]: ...on further checking, maybe they are not statically linked - beyond my understanding - but they are much larger than the binaries in the distros

File: docs/root/start/_include/dockerhub-images.csv (hunk 5 above)
Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:9]: > Ive tried to put info about all active images in one place - to be authoritative.

that said, i also think if we do keep it here we need to make clear it doesnt contain  an envoy binary, is very large etc - perhaps putting in a separate table.

File: docs/root/start/_include/dockerhub-images.csv (hunk 7 above)
Comment (mattklein123) [reply docs/root/start/_include/dockerhub-images.csv:3]: OK I don't really have a strong opinion. In general I don't want to let perfect be the enemy of the good. This is a massive improvement so we can ship and iterate.

File: docs/root/start/_include/dockerhub-images.csv (hunk 5 above)
Comment (mattklein123) [reply docs/root/start/_include/dockerhub-images.csv:9]: In terms of what to do with the release versions and the SHA if the build image, I think this can very easily be grammatically generated as part of `conf.py` in the sphinx build. The version can be inferred from: https://github.com/envoyproxy/envoy/blob/master/VERSION (you can even have the table generated different depending on whether it's release docs or not, up to you). The build SHA can be pulled from: https://github.com/envoyproxy/envoy/blob/master/ci/envoy_build_sha.sh.

My advice is to just get something working and we can bikeshed about the "optimal" flow later.

File: docs/root/start/building/local_docker_build.rst (hunk 9 above)
Comment (mattklein123) [reply docs/root/start/building/local_docker_build.rst:23]: As I said elsewhere, I would rather not let perfect be the enemy of the good. Can we land this overall improvement and then we can discuss what other changes you want to make? 

Personally, I think you are overstimating how important building Envoy is to the average user and I would rather it be put in some other section but we can sort that out later.

**Comment by mattklein123** [comment]
LGTM to ship and iterate modulo main merge and possibly programmatically generating the various versions and SHAs.

/wait

File: docs/root/start/_include/dockerhub-images.csv (hunk 5 above)
Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:9]: on the basis that generating this csv file is outside the scope of this PR i have put "see dockerhub" here

File: docs/root/start/building/local_docker_build.rst (hunk 9 above)
Comment (phlax) [reply docs/root/start/building/local_docker_build.rst:23]: ive strengthened the warning on the first command.

the issue with the second command should be resolved by #13618

File: docs/root/start/_include/dockerhub-images.csv (hunk 7 above)
Comment (phlax) [reply docs/root/start/_include/dockerhub-images.csv:3]: ive readded the not about `X.Y-latest` versions and some info about the build image in the note under the table

**Comment by mattklein123** [comment]
Awesome, suggest we ship and iterate from here. Can you merge main?

/wait

**Comment by phlax** [comment]
cool, yep, just did, and fixing redirects

**Comment by phlax** [comment]
@mattklein123 i think we need to wait for tests

**Comment by mattklein123** [comment]
> @mattklein123 i think we need to wait for tests

SG. Also, friendly request to please not force push. It makes reviews a lot more difficult.

**Review by mattklein123 — APPROVED** [review]

**Comment by phlax** [comment]
> SG. Also, friendly request to please not force push. It makes reviews a lot more difficult.

apologies, im more used to doing that to resolve confllicts. ill try to rem not  to

on a related not see #13735

**Comment by phlax** [comment]
aaggh - #13735 is causing master fail

**Review by mattklein123 — APPROVED** [review]
